desc/objectives/_coils.py

```diff
@@ -597,3 +598,274 @@ def compute(self, params, constants=None):
+        the optimization of the ``field`` object.
+    external_field_source_grid : Grid, optional
+        Grid object used to discretize the external field source. Defaults to the
+        default for the given external_field, see the docstring of the field object
+        for the specific default.
+    name : str, optional
+        Name of the objective function.
+    """
+
+    _coordinates = "rtz"
+    _units = "T"
+    _print_value_fmt = "Quadratic Flux: {:10.3e} "
+
+    def __init__(
+        self,
+        eq,
+        field,
+        target=None,
+        bounds=None,
+        weight=1,
+        normalize=True,
+        normalize_target=True,
+        loss_function=None,
+        deriv_mode="auto",
+        source_grid=None,
+        eval_grid=None,
+        external_field=None,
+        external_field_source_grid=None,
+        name="toroidal-flux",
+        eq_fixed=False,
+    ):
+        if target is None and bounds is None:
+            target = eq.Psi
+        self._field = field
+        self._source_grid = source_grid
+        self._eval_grid = eval_grid
+        self._eq_fixed = eq_fixed
+        self._eq = eq if eq_fixed else None
+        self._external_field = external_field
+        self._external_field_source_grid = external_field_source_grid
+
+        super().__init__(
+            things=[field, eq] if not eq_fixed else [field],
+            target=target,
+            bounds=bounds,
+            weight=weight,
+            normalize=normalize,
+            normalize_target=normalize_target,
+            loss_function=loss_function,
+            deriv_mode=deriv_mode,
+            name=name,
+        )
+
+    def build(self, use_jit=True, verbose=1):
+        """Build constant arrays.
+
+        Parameters
+        ----------
+        use_jit : bool, optional
+            Whether to just-in-time compile the objective and derivatives.
+        verbose : int, optional
+            Level of output.
+
+        """
+        eq = self._eq if self._eq_fixed else self.things[1]
+        field = self.things[0]
+        # if field is different than self._field, update
+        if field != self._field:
+            self._field = field
+        # if eq is different than self._eq, update
+        if eq != self._eq:
+            self._eq = eq
+        if self._eval_grid is None:
+            eval_grid = LinearGrid(
+                L=eq.L_grid, M=eq.M_grid, zeta=jnp.array(0.0), NFP=eq.NFP
+            )
+            self._eval_grid = eval_grid
+        else:
+            eval_grid = self._eval_grid
+
+        warnif(
+            not np.allclose(eval_grid.nodes[:, 2], eval_grid.nodes[0, 2]),
+            UserWarning,
+            "Evaluation grid should be at constant zeta",
+        )
+
+        # ensure vacuum eq, as we don't yet support finite beta
+        pres = np.max(np.abs(eq.compute("p")["p"]))
+        curr = np.max(np.abs(eq.compute("current")["current"]))
+        warnif(
```

Comment:
would it be better to error out here, since the results might be meaningless if there is a plasma field?

Also, might be clearer to say that this objective isn't needed at finite beta